### PR TITLE
RavenDB-14250 Allow to fix database topology

### PR DESCRIPTION
**Pull request by karmeli87**
With this change the order of the members is fixed.
Note that as a side-effect moving from member to rehab is disabled.

### CODE REVIEW

**Review by ayende — COMMENTED** [review]
I don't think that this is a complete situation.
Consider the case of A,B,C topology, where A is the primary.
It goes down, and then we have B,C as members, and A as rehab.
But once A is back online, we want it to be the primary again, not sit there in the end.

So the `Fixed` shouldn't be just a boolean, but rather be something like `DesiredTopology` or something like that, so users can say how they want things to show up.

**Comment by grisha-kotler** [comment]
Do we need to change anything in the studio?

**Comment by karmeli87** [comment]
Yes, when re-ordering database nodes we need to add an option to "fix" this order.

**Comment by arekpalinski** [comment]
Will you need @ml054 to take care of it?

File: src/Raven.Client/ServerWide/DatabaseTopology.cs [hunk 1]
```diff
@@ -153,6 +153,40 @@ public class DatabaseTopology
         public LeaderStamp Stamp;
         public bool DynamicNodesDistribution;
         public int ReplicationFactor = 1;
+        public List<string> PriorityOrder;
+
+        private readonly Random _random = new Random();
```
Comment (ayende) [inline src/Raven.Client/ServerWide/DatabaseTopology.cs]: We only need it for a specific scenario, this shouldn't be here.

File: src/Raven.Client/ServerWide/DatabaseTopology.cs [hunk 2]
```diff
@@ -153,6 +153,40 @@ public class DatabaseTopology
...
+        public int Shuffle(string _, string __)
```
Comment (ayende) [inline src/Raven.Client/ServerWide/DatabaseTopology.cs]: Same, shouldn't be on this class, certainly not as a public method.
Create a separate class that can use this

File: src/Raven.Client/ServerWide/DatabaseTopology.cs [hunk 3]
```diff
@@ -153,6 +153,40 @@ public class DatabaseTopology
...
+
+            var members = new List<string>();
+            for (int i = 0; i < PriorityOrder.Count; i++)
+            {
+                var priorityNode = PriorityOrder[i];
+                if (Members.Contains(priorityNode))
+                    members.Add(priorityNode);
+            }
+
+            for (int i = 0; i < Members.Count; i++)
```
Comment (ayende) [inline src/Raven.Client/ServerWide/DatabaseTopology.cs:184]: Should we shuffle the other stuff, that isn't in the priority order?

Comment (karmeli87) [reply src/Raven.Client/ServerWide/DatabaseTopology.cs:184]: okay, so I shuffle the members any way

File: src/Raven.Client/ServerWide/DatabaseTopology.cs (hunk 2 above)
Comment (karmeli87) [reply src/Raven.Client/ServerWide/DatabaseTopology.cs]: done

File: src/Raven.Client/ServerWide/DatabaseTopology.cs (hunk 1 above)
Comment (karmeli87) [reply src/Raven.Client/ServerWide/DatabaseTopology.cs]: done

File: src/Raven.Client/ServerWide/DatabaseTopology.cs [hunk 4]
```diff
@@ -140,6 +140,17 @@ public override int GetHashCode()
         }
     }
 
+    public static class ThreadSafeRandom
```
Comment (arekpalinski) [inline src/Raven.Client/ServerWide/DatabaseTopology.cs]: Can we change it to `private` or `internal`? I don't want it to be considered public API

File: src/Raven.Client/ServerWide/DatabaseTopology.cs [hunk 5]
```diff
@@ -153,6 +164,33 @@ public class DatabaseTopology
         public LeaderStamp Stamp;
         public bool DynamicNodesDistribution;
         public int ReplicationFactor = 1;
+        public List<string> PriorityOrder;
+
+        public void ReorderMembers()
```
Comment (arekpalinski) [inline src/Raven.Client/ServerWide/DatabaseTopology.cs]: Can we change it to `internal`?

File: src/Raven.Client/ServerWide/Operations/ReorderDatabaseMembersOperation.cs [hunk 6]
```diff
@@ -15,20 +15,22 @@ public class ReorderDatabaseMembersOperation : IServerOperation
...
         {
             public List<string> MembersOrder;
+            public bool Fixed;
         }
 
         private readonly string _database;
         private readonly Parameters _parameters;
 
-        public ReorderDatabaseMembersOperation(string database, List<string> order)
+        public ReorderDatabaseMembersOperation(string database, List<string> order, bool fixedTopology = false)
```
Comment (arekpalinski) [inline src/Raven.Client/ServerWide/Operations/ReorderDatabaseMembersOperation.cs]: This breaks binary compatibility. Please add a new constructor for that overload.

File: src/Raven.Client/ServerWide/DatabaseTopology.cs (hunk 4 above)
Comment (karmeli87) [reply src/Raven.Client/ServerWide/DatabaseTopology.cs]: done

File: src/Raven.Client/ServerWide/DatabaseTopology.cs (hunk 5 above)
Comment (karmeli87) [reply src/Raven.Client/ServerWide/DatabaseTopology.cs]: done

File: src/Raven.Client/ServerWide/Operations/ReorderDatabaseMembersOperation.cs (hunk 6 above)
Comment (karmeli87) [reply src/Raven.Client/ServerWide/Operations/ReorderDatabaseMembersOperation.cs]: done

**Pull request merged**